### PR TITLE
Generalize action testing framework

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Generalize action testing framework to:

- Have a local algorithm hold the DataBox (necessary for Actions that call a simple action)
- Replace some uses of `apply` with a function called `next_action` that evaluates only the next action in the action list. If the action is not ready and `ERROR` is hit.
- Replace remaining uses of `apply` with a function call `simple_action` that invokes the simple action on the component.

**Note:** Only commits the following commits are new:
- Replace apply with simple_action in mocking framework
- Add next_action, is_ready support to mocking framework
- Change returned DataBox type to rvalue in EventsAndTriggers
- Have mocking framework create all local algorithms
- Allow actions to retrieve component mocks
- Compute DataBox types in mocking framework
- Enclose ActionTesting_detail in ActionTesting namespace
- Move code around ActionRunner class

Depends on:
- [x] #912 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+// IWYU pragma: no_include "DataStructures/VariablesHelpers.hpp"  // for Variables
+
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp]: Remove blank line.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+#include "NumericalAlgorithms/DiscontinuousGalerkin/Tags.hpp"
+#include "NumericalAlgorithms/Spectral/Projection.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+// IWYU pragma: no_include "Parallel/PupStlCpp11.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/StdHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "tests/Unit/ActionTesting.hpp"
+// #include "tests/Unit/TestHelpers.hpp"
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp]: Remove?

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+          ActionTesting::MockLocalAlgorithm<my_component>{db::create<
+              db::AddSimpleTags<
+                  TemporalId, Tags::Next<TemporalId>, Tags::Mesh<2>,
+                  Tags::Element<2>, Tags::ElementMap<2>,
+                  normal_dot_fluxes_tag<2, flux_comm_types<2>>,
+                  other_data_tag<2>, mortar_meshes_tag<2>, mortar_sizes_tag<2>,
+                  MortarRecorderTag, mortar_next_temporal_ids_tag<2>>,
+              typename my_component::compute_tags>(
+              self_step_start, self_step_end,
+
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp]: Remove blank line.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp [hunk 4]
```diff
@@ -434,51 +429,31 @@ SPECTRE_TEST_CASE("Unit.DiscontinuousGalerkin.Actions.FluxCommunication",
...
-                                                                     self_id));
+  CHECK_FALSE(runner.is_ready<my_component>(self_id));
+  runner.next_action<my_component>(south_id);
+  CHECK_FALSE(runner.is_ready<my_component>(self_id));
+  runner.next_action<my_component>(east_id);
+  CHECK_FALSE(runner.is_ready<my_component>(self_id));
+  runner.next_action<my_component>(west_id);
+  CHECK(runner.is_ready<my_component>(self_id));
+
+  // DeceiveDataForFluxes
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp]: I know we're mocking, but...

**Review by wthrowe — CHANGES_REQUESTED** [review]
Travis OSX failures are real.

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp (hunk 4 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunication.cpp]: I really enjoyed this typo, thank you for noticing it :)

File: tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp (hunk 2 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Actions/Test_FluxCommunicationLts.cpp]: Yea, IWYU experiment, forgot to delete...

**Comment by nilsdeppe** [comment]
I've done the following:
- Rebased onto the other PR (I'll do it again once that one gets merged)
- Added fixups addressing comments (thanks for catching all those!)
- Added a commit to allow better interoperability with libc++ (I was trying to test the macOS failure on my machine, hoping my libc++ version was close enough to reproduce, it wasn't)
- I've squashed a commit that hopefully fixes the macOS build errors. I believe they were due to libc++ not explicitly converting the ElementId to ElementIndex in the emplace call.

**Comment by nilsdeppe** [comment]
Rebased on develop (@wthrowe can you verify the changes to Test_InitializeElement again? I want to make sure I didn't break anything that tests LTS :) )

**Review by kidder — DISMISSED** [review]
Don't be surprised if I make suggestions after trying to use this framework...

**Comment by nilsdeppe** [comment]
I'll be disappointed if you don't at least make documentation suggestions ;)

**Review by wthrowe — DISMISSED** [review]
Test_InitializeElement looks fine.

**Comment by nilsdeppe** [comment]
Okay, thanks for checking that again!

Squashed the commits.

**Review by kidder — APPROVED** [review]

**Comment by nilsdeppe** [comment]
Travis failures are timeouts

**Comment by kidder** [comment]
@wthrowe I think this is ready

**Review by wthrowe — APPROVED** [review]

**Pull request merged**